py/kubeflow/kfctl/testing/util/gcp_util.py

```diff
@@ -122,46 +122,84 @@ def iap_is_ready(url, wait_min=15):
     sleep(10)
   return False
 
-def basic_auth_is_ready(url, username, password, wait_min=15):
-  get_url = url + "/kflogin"
-  post_url = url + "/apikflogin"
+def _send_req(wait_sec, url, req_gen):
+  """ Helper function to send requests and retry when the endpoint is not ready.
 
-  req_num = 0
-  end_time = datetime.datetime.now() + datetime.timedelta(
-      minutes=wait_min)
-  while datetime.datetime.now() < end_time:
-    req_num += 1
-    logging.info("Trying url: %s request number %s" % (get_url, req_num))
+  Args:
+  wait_sec: int, max time to wait and retry in seconds.
+  url: str, url to send the request, used only for logging.
+  req_gen: lambda, no parameter function to generate requests.Request for the
+  function to send to the endpoint.
+
+  Returns:
+    requests.Response
+  """
+
+  # Simple identifier that we need to re-send the request.
+  class RetryError(Exception):
+    pass
+
+  def retry_if_retry_error(e):
+    return isinstance(e, RetryError)
+
+  @retry(stop_max_delay=wait_sec * 1000, wait_fixed=10 * 1000,
+         retry_on_exception=retry_if_retry_error)
+  def _send(url, req_gen):
     resp = None
+    logging.info("sending request to %s" % url)
     try:
-      resp = requests.request(
-          "GET",
-          get_url,
-          verify=False)
+      resp = req_gen()
     except SSLError as e:
-      logging.warning("%s: Endpoint SSL handshake error: %s; request number: %s" % (url, e, req_num))
+      logging.warning("%s: Endpoint SSL handshake error: %s" % (url, e))
```

Comment:
Why convert SSLError into RetryError? Why not just have the retry_on_exception test if its an SSLError or ReqConnectionError?

Reply:
updated.